src/Sylius/Bundle/ApiBundle/StateProvider/Shop/Order/Payment/PaymentMethod/CollectionProvider.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Sylius Sp. z o.o.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ApiBundle\StateProvider\Shop\Order\Payment\PaymentMethod;
+
+use ApiPlatform\Metadata\GetCollection;
+use ApiPlatform\Metadata\Operation;
+use ApiPlatform\State\ProviderInterface;
+use Doctrine\Common\Collections\Collection;
+use Sylius\Bundle\ApiBundle\SectionResolver\ShopApiSection;
+use Sylius\Bundle\ApiBundle\Serializer\ContextKeys;
+use Sylius\Bundle\CoreBundle\SectionResolver\SectionProviderInterface;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Core\Model\PaymentInterface;
+use Sylius\Component\Core\Model\PaymentMethodInterface;
+use Sylius\Component\Core\Repository\PaymentRepositoryInterface;
+use Sylius\Component\Payment\Resolver\PaymentMethodsResolverInterface;
+use Webmozart\Assert\Assert;
+
+/** @implements ProviderInterface<PaymentMethodInterface> */
+final readonly class CollectionProvider implements ProviderInterface
+{
+    public function __construct(
+        private PaymentRepositoryInterface $paymentRepository,
+        private SectionProviderInterface $sectionProvider,
+        private PaymentMethodsResolverInterface $paymentMethodsResolver,
+    ) {
+    }
+
+    public function provide(Operation $operation, array $uriVariables = [], array $context = []): array|Collection
+    {
+        Assert::true(is_a($operation->getClass(), PaymentMethodInterface::class, true));
+        Assert::isInstanceOf($operation, GetCollection::class);
+        Assert::isInstanceOf($this->sectionProvider->getSection(), ShopApiSection::class);
+        Assert::keyExists($context, ContextKeys::CHANNEL);
+        Assert::keyExists($uriVariables, 'tokenValue');
+        Assert::keyExists($uriVariables, 'paymentId');
+
+        /** @var ChannelInterface $channel */
+        $channel = $context[ContextKeys::CHANNEL];
+
+        /** @var PaymentInterface|null $shipment */
+        $payment = $this->paymentRepository->findOneByOrderTokenAndChannel(
+            $uriVariables['paymentId'],
+            $uriVariables['tokenValue'],
+            $channel,
+        );
```

Comment:
Although this might be faster, I feel like debugging the case when the current channel !== order's channel will be a chore, since the only info we have here is the payment or lack thereof.